trace_replay/block_cache_tracer.cc

```diff
@@ -215,4 +212,47 @@ Status BlockCacheTraceReader::ReadAccess(BlockCacheTraceRecord* record) {
   return Status::OK();
 }
 
+AtomicBlockCacheTraceWriter::AtomicBlockCacheTraceWriter() {
+  writer_.store(nullptr);
+}
+
+AtomicBlockCacheTraceWriter::~AtomicBlockCacheTraceWriter() { EndTrace(); }
+
+Status AtomicBlockCacheTraceWriter::StartTrace(
+    Env* env, const TraceOptions& trace_options,
+    std::unique_ptr<TraceWriter>&& trace_writer) {
+  InstrumentedMutexLock lock_guard(&writer_mutext_);
```

Comment:
I thought you want to check the value of `writer_` before acquiring the mutex?

Reply:
I check the value of writer_ in WriteBlockAccess before acquiring the mutex. If writer_ is null, WriteBlockAccess is a noop.  

For StartTrace and EndTrace, I acquire the lock in the beginning to make the code clearer as it avoids race conditions with concurrent threads calling StartTrace or EndTrace.